.github/workflows/ci.yaml

```
@@ -0,0 +1,103 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: apptainer charm tests
+on:
+  workflow_call:
+  pull_request:
+
+jobs:
+  inclusive-naming:
+    name: Inclusive naming checks
+    runs-on: ubuntu-24.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Install `just`
+        run: sudo snap install just --classic
+      - name: Install `woke`
+        run: sudo snap install woke
+      - name: Run inclusive naming checks
+        run: just woke
+
+  lint:
+    name: Lint checks
+    runs-on: ubuntu-24.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Install `just`
+        run: sudo snap install just --classic
+      - name: Install `uv`
+        run: sudo snap install astral-uv --classic
+      - name: Run lint checks
+        run: just lint
+      - name: Run lint checks
+        run: just lint
+
+  typecheck:
+    name: Static type checks
+    runs-on: ubuntu-24.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Install `just`
+        run: sudo snap install just --classic
+      - name: Install `uv`
+        run: sudo snap install astral-uv --classic
+      - name: Run static type checks
+        run: just typecheck
+
+  unit-test:
+    name: Unit tests
+    runs-on: ubuntu-24.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Install `just`
+        run: sudo snap install just --classic
+      - name: Install `uv`
+        run: sudo snap install astral-uv --classic
+      - name: Run unit tests
+        run: just unit
+
+  integration-test:
+    name: Integration tests (LXD)
+    runs-on: ubuntu-24.04
+    needs:
+      - lint
+      - typecheck
+      - unit-test
+    steps:
+      - name: Remove unnecessary files
+        run: |
+          sudo rm -rf /usr/share/dotnet
+          sudo rm -rf /opt/ghc
+          sudo rm -rf /usr/local/share/boost
+          sudo rm -rf "$AGENT_TOOLSDIRECTORY"
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Install `just`
+        run: sudo snap install just --classic
+      - name: Install `uv`
+        run: sudo snap install astral-uv --classic
+      - name: Install `charmcraft`
+        run: sudo snap install charmcraft --classic
+      - name: Setup operator environment
+        uses: charmed-kubernetes/actions-operator@main
+        with:
+          provider: lxd
+          juju-channel: 3.6/stable
+      - name: Run integration tests
+        run: just integration --base ubuntu@24.04
```

```suggestion
        run: just integration --base ubuntu@24.04

```
Missing last newline

---

Extraneous whitespace removed in commit `436a589`